sdk/cosmos/azure-cosmos-spark_3-4_2-12/src/main/scala/com/azure/cosmos/spark/CosmosRowConverter.scala

```diff
@@ -0,0 +1,443 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.spark.SchemaConversionModes.SchemaConversionMode
+import com.fasterxml.jackson.annotation.JsonInclude.Include
+// scalastyle:off underscore.import
+import com.fasterxml.jackson.databind.node._
+import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
+import org.apache.spark.sql.catalyst.expressions.{UnsafeMapData}
+import org.apache.spark.sql.catalyst.util.ArrayData
+
+import java.sql.{Date, Timestamp}
+import java.time.format.DateTimeFormatter
+import java.time.{Instant, LocalDate, LocalDateTime, ZoneOffset}
+import java.util.concurrent.TimeUnit
+import scala.collection.concurrent.TrieMap
+
+// scalastyle:off underscore.import
+import org.apache.spark.sql.types._
+import scala.collection.JavaConverters._
+// scalastyle:on underscore.import
+
+import scala.util.{Try, Success, Failure}
+
+// scalastyle:off
+private[cosmos] object CosmosRowConverter {
+
+  // TODO: Expose configuration to handle duplicate fields
+  // See: https://github.com/Azure/azure-sdk-for-java/pull/18642#discussion_r558638474
+  private val rowConverterMap = new TrieMap[CosmosSerializationConfig, CosmosRowConverter]
+
+  def get(serializationConfig: CosmosSerializationConfig) : CosmosRowConverter = {
+    rowConverterMap.get(serializationConfig) match {
+      case Some(existingRowConverter) => existingRowConverter
+      case None =>
+        val newRowConverterCandidate = createRowConverter(serializationConfig)
+        rowConverterMap.putIfAbsent(serializationConfig, newRowConverterCandidate) match {
+          case Some(existingConcurrentlyCreatedRowConverter) => existingConcurrentlyCreatedRowConverter
+          case None => newRowConverterCandidate
+        }
+    }
+  }
+
+  private def createRowConverter(serializationConfig: CosmosSerializationConfig): CosmosRowConverter = {
+    val objectMapper = new ObjectMapper()
+      import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
+      objectMapper.registerModule(new JavaTimeModule)
+    serializationConfig.serializationInclusionMode match {
+      case SerializationInclusionModes.NonNull => objectMapper.setSerializationInclusion(Include.NON_NULL)
+      case SerializationInclusionModes.NonEmpty => objectMapper.setSerializationInclusion(Include.NON_EMPTY)
+      case SerializationInclusionModes.NonDefault => objectMapper.setSerializationInclusion(Include.NON_DEFAULT)
+      case _ => objectMapper.setSerializationInclusion(Include.ALWAYS)
+    }
+
+    new CosmosRowConverter(objectMapper, serializationConfig)
+  }
+}
+
+private[cosmos] class CosmosRowConverter(private val objectMapper: ObjectMapper, private val serializationConfig: CosmosSerializationConfig)
```

Comment:
I did this way because TimestampNTZType is a new data type, so of course it can't be added to base class at all (does not exist prior to 3.4). I have only overriden the methods that have match cases for each possible data type, in order to be able to add a case for the new data type (TimestampNTZType), though of course this does mean a lot of duplication: 

convertSparkDataTypeToJsonNodeConditionally 
convertSparkDataTypeToJsonNodeNonNull
convertToSparkDataType

The other methods are custom methods for the TimestampNTZType conversion. I did not see a more parsimonious way of doing it?